mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -1518,24 +1511,192 @@ def _log_inputs_impl(
+
+        Args:
+            request_id: Unique string identifier of the trace.
+
+        Returns:
+            The TraceInfo object.
+        """
+        with self.ManagedSessionMaker() as session:
+            sql_trace_info = self._get_sql_trace_info(session, request_id)
+            return sql_trace_info.to_mlflow_entity()
+
+    def _generate_trace_request_id(self) -> str:
+        return uuid.uuid4().hex
+
+    def _get_sql_trace_info(self, session, request_id) -> SqlTraceInfo:
+        sql_trace_info = (
+            session.query(SqlTraceInfo).filter(SqlTraceInfo.request_id == request_id).one_or_none()
+        )
+        if sql_trace_info is None:
+            raise MlflowException(
+                f"Trace with request_id '{request_id}' not found.", RESOURCE_DOES_NOT_EXIST
+            )
+        return sql_trace_info
+
+    def search_traces(
+        self,
+        experiment_ids: List[str],
+        filter_string: Optional[str] = None,
+        max_results: int = SEARCH_TRACES_DEFAULT_MAX_RESULTS,
+        order_by: Optional[List[str]] = None,
+        page_token: Optional[str] = None,
+    ) -> Tuple[List[TraceInfo], Optional[str]]:
+        """
+        Return traces that match the given list of search expressions within the experiments.
+
+        Args:
+            experiment_ids: List of experiment ids to scope the search.
+            filter_string: A search filter string.
+            max_results: Maximum number of traces desired.
+            order_by: List of order_by clauses.
+            page_token: Token specifying the next page of results. It should be obtained from
+                a ``search_traces`` call.
+
+        Returns:
+            A tuple of a list of :py:class:`TraceInfo <mlflow.entities.TraceInfo>` objects that
+            satisfy the search expressions and a pagination token for the next page of results.
+        """
+        self._validate_max_results_param(max_results)
+
+        with self.ManagedSessionMaker() as session:
+            cases_orderby, parsed_orderby, sorting_joins = _get_orderby_clauses_for_search_traces(
+                order_by or [], session
+            )
+            stmt = select(SqlTraceInfo, *cases_orderby)
+
+            attribute_filters, non_attribute_filters = _get_filter_clauses_for_search_traces(
+                filter_string, session, self._get_dialect()
+            )
+            for non_attr_filter in non_attribute_filters:
+                stmt = stmt.join(non_attr_filter)
+
+            # using an outer join is necessary here because we want to be able to sort
+            # on a column (tag, metric or param) without removing the lines that
+            # do not have a value for this column (which is what inner join would do)
+            for j in sorting_joins:
+                stmt = stmt.outerjoin(j)
+
+            offset = SearchTraceUtils.parse_start_offset_from_page_token(page_token)
+            stmt = (
+                stmt.distinct()
```

Comment:
Do we need the distinct() call here? This is PKEY == PKEY for the left outer join, right? This might dramatically improve performance if we omit this full post-join table scan

Reply:
Ah that seems to be true!

Indeed, the `request_id` is not primary key on the right table, which is in this case either `TraceTag` or `TraceRequestMetadata` table. Those tables originally can contain multiple rows like `("tr-12345", "key1", "val1"), ("tr-12345", "key2", "val2")`.

However, the trick here is that the subquery conditioned by tag/metadata value always ensures the uniqueness to `request_id`. This is consequence of the facts that we don't support `OR` syntax and construct a separate subquery for each tag condition. For example, if we have filter string `tag.x = 'foo' and tag.y = 'bar'`, then we construct two subqueries where each of them conditions on each key. The tag/metadata tables guarantees uniqueness for the combination of `request_id` and `key`, so the subquery output (filtered with a single key), should not have duplicated rows for `request_id`.

This behavior will break when we start to have `OR` clause **and** consolidate multiple subqueries into one (which may improve performance in some engines). But at this moment, we don't require the deduplication. I will remove that and also add a comment explaining the risk of duplicated records if we modify the query building logic🙂